iabtcf-encoder/src/main/java/com/iabtcf/encoder/BitFieldEncoder.java

```diff
@@ -0,0 +1,166 @@
+        this.vendors = vendors;
+        this.maxVendorId = maxVendorId;
+    }
+
+    public BitFieldEncoder add(int vendorId) {
+        vendors.set(vendorId - 1);
+
+        return this;
+    }
+
+    public BitFieldEncoder add(Iterable<Integer> vendorIds) {
+        for (Iterator<Integer> i = vendorIds.iterator(); i.hasNext();) {
+            add(i.next());
+        }
+
+        return this;
+    }
+
+    public BitFieldEncoder add(int... vendorIds) {
+        for (int i = 0; i < vendorIds.length; i++) {
+            add(vendorIds[i]);
+        }
+
+        return this;
+    }
+
+    public BitFieldEncoder add(IntIterable ii) {
+        for (IntIterator i = ii.intIterator(); i.hasNext();) {
+            add(i.nextInt());
+        }
+
+        return this;
+    }
+
+    /**
+     * Emit the specified max vendor id. By default, the maximum vendorId that was added is emitted.
+     */
+    public BitFieldEncoder setMaxVendorId(int maxVendorId) {
+        this.maxVendorId = maxVendorId;
+
+        return this;
+    }
+
+    /**
+     * Returns a BitWriter that represents the encoded vendor ids. The BitWriter may either encode a
+     * bit field or a range encoding; depending on which is smaller.
+     */
+    public BitWriter build() {
+        BitWriter bv = new BitWriter();
+
+        if (vendors.length() == 0) {
+            bv.write(0, FieldDefs.CORE_VENDOR_MAX_VENDOR_ID);
+            bv.write(false, FieldDefs.CORE_VENDOR_IS_RANGE_ENCODING);
+            return bv;
+        }
+
+        maxVendorId = maxVendorId == -1 ? vendors.length() : maxVendorId;
+
+        // create the range bit section
+        BitWriter rangeBits = new BitWriter();
+        int idxSet = vendors.get(0) ? 0 : vendors.nextSetBit(0);
+        int idxClr;
+        int numEntries = 0;
+        do {
+            idxClr = vendors.nextClearBit(idxSet);
+            int length = idxClr - idxSet;
+
+            if (length == 1) {
+                rangeBits.write(false, FieldDefs.CORE_VENDOR_IS_RANGE_ENCODING);
+                rangeBits.write(idxSet + 1, FieldDefs.START_OR_ONLY_VENDOR_ID);
+            } else {
+                rangeBits.write(true, FieldDefs.CORE_VENDOR_IS_RANGE_ENCODING);
+                rangeBits.write(idxSet + 1, FieldDefs.START_OR_ONLY_VENDOR_ID);
+                rangeBits.write(length, FieldDefs.END_VENDOR_ID);
+            }
+            numEntries++;
+        } while ((idxSet = vendors.nextSetBit(idxClr)) > 0 && rangeBits.length() < vendors.length());
+
+        // emit max vendor id
+        bv.write(maxVendorId, FieldDefs.CORE_VENDOR_MAX_VENDOR_ID);
+
+        if (rangeBits.length() < vendors.length()) {
+            // emit range bits
+            bv.write(true, FieldDefs.IS_A_RANGE);
+
+            bv.write(numEntries, FieldDefs.NUM_ENTRIES);
+            bv.write(rangeBits);
+        } else {
+            // emit bit field
+            bv.write(false, FieldDefs.IS_A_RANGE);
+
+            byte[] bits = vendors.toByteArray();
```

Comment:
I don't understand why you need all this `byte` stuff???  I would think that you would just do:

```Java
for(int i = 0; i < maxVendorId; i++) {
  bv.write(vendors.get(i));
}
```

Reply:
yeah we can do that